### ISSUE TITLE
Custom text field in dataset array doesn't work (templateResult callback).

### ISSUE THREAD

**Issue by Roquie** [select2/select2#4266]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of Select2
- [x] This is not a usage question (Those should be directed to the [community](https://select2.github.io/community.html))
- [x] I have attempted to find the simplest possible steos to reproduce the issue

## Steps to reproduce the issue

1.First, code example:
```js
var data = [{ id: 0, name: 'enhancement' }, { id: 1, name: 'bug' }, { id: 2, name: 'duplicate' }, { id: 3, name: 'invalid' }, { id: 4, name: 'wontfix' }];
$('select').select2({
   data: data,
   templateResult: function (result, container) {
         return result.name;
   },
   templateSelection: function (selection) {
         return selection.name;
   }
});
```
2.Open the browser console. So, we see that `templateSelection` will be displayed normally. But, `templateResult` applying empty strings to each `select` option.
![2016-03-23 19 50 03](https://cloud.githubusercontent.com/assets/3214290/13993204/694c0776-f130-11e5-9d38-a53457888d11.png)

3.Results:
```js
$('select').select2().val(1).trigger('change');
```
![2016-03-23 20 40 00](https://cloud.githubusercontent.com/assets/3214290/13994763/5dc47f3a-f137-11e5-9fd2-9c531f8e4b27.png)

## Environment

Libraries

- jQuery version: 3.0-beta-1
- Select2 version: 4.0.2/master

## Bugfix

1. Bug placed in [link](https://github.com/select2/select2/blob/master/src/js/select2/data/select.js#L162)
2. Fix
```js
SelectAdapter.prototype.option = function (data) {
    var option;
    var template = this.options.get('templateResult');
    var content = template(data); // container variable not passed.

    if (data.children) {
      option = document.createElement('optgroup');
      option.label = content;
    } else {
      option = document.createElement('option');

      if (option.textContent !== undefined) {
        option.textContent = content;
      } else {
        option.innerText = content;
      }
    }
//...etc
```
...but, I am [not sure](https://github.com/select2/select2/blob/master/src/js/select2/results.js#L500) in correct this solution (`container` not passed to second argument).

## Workaround

```js

$('select').select2({
   data: $.map(data, function (item) {
       return {
          id: item.id,
          text: item.name
      };
   })
});
```

**Comment by kevin-brown** [comment]
The `templateResult` option does not control what field is used to populate the `<option>` tag later, that's still the `text` field.

https://select2.github.io/options.html#my-objects-use-a-property-other-than-text-for-the-text-that-nee

It's just supposed to be used to render additional information, or using different styling. It is not supposed to be used to rename the `text` property, there are better ways to do that.

As far as 3 goes, the issue there is that you are calling `select2()` which will reinitialize Select2 with the default configuration. This will remove your `templateResult` and `templateSelection` callbacks, which doesn't appear to be what you were expecting.

**Comment by Roquie** [comment]
Hmmm ... okay.
But, don't you're think it is overhead? Every time use map function? Why not use two simple callbacks (concept):
```js
$('select').select2({
   data: function (response) {
       return response.data; //or by default data is `response` or `response.data`;
   },
   processItem: function (item) {
      return {
            id: item.id,
            text: item.name // item.name or item.description or item.content or ...
            // Integration with different api's often require different names.
      };
   }
});
```

**Comment by kevin-brown** [comment]
> But, don't you're think it is overhead? Every time use map function? Why not use two simple callbacks

There is going to be overhead either way, the question is where do we want to put it.

I'm trying to avoid adding the overhead and maintenance burden to Select2, especially when it can be solved outside of Select2 for the cases which require it. It _was previously_ done within Select2, but we eventually determined that it would be better to standardize the properties which we require and make it easy for people who are required to use a standard format to work with the format which Select2 uses.

**Comment by Roquie** [comment]
> It **was previously done** within Select2, but we eventually determined that it would be better to standardize the properties which we require and make it easy for people who are required to use a standard format to work with the format which Select2 uses.

What was the problem? How solution on top prevents this?

**Comment by kevin-brown** [comment]
> What was the problem?

The core problem comes down to the fact that it added an additional maintenance cost to Select2. Anywhere that Select2 referred to the identifiers or textual representations (`id` and `text` properties), we had to ensure that we called whatever `id` or `text` callback existed. This meant that anyone modifying Select2 had to be aware of that, and over time pull requests were accepted which didn't consider it. This resulted in features that were broken when the callbacks were used, but not always consistently.

> How solution on top prevents this?

By standardizing the object interface, Select2 no longer has to handle the special case of people not using `id` and `text` for their properties. Special cases add more to the maintenance burden of Select2, which is something we always aim to keep low. Considering 90% of people can use `id` and `text` for their properties, we leave the property mapping to the other 10%.
